core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/AbstractFederationConnection.java

```diff
@@ -434,10 +439,27 @@ private TupleExpr optimize(TupleExpr parsed, Dataset dataset, BindingSet binding
 		new QueryMultiJoinOptimizer().optimize(query, dataset, bindings);
 		// new FilterOptimizer().optimize(query, dataset, bindings);
 
-		new EmptyPatternOptimizer(members).optimize(query, dataset, bindings);
+		// prepare bloom filters
+		RepositoryBloomFilter defaultBloomFilter = new AccurateRepositoryBloomFilter(includeInferred);
+		Map<Repository, RepositoryBloomFilter> bloomFilters = federation.getBloomFilters();
+		Map<RepositoryConnection, RepositoryBloomFilter> bloomFiltersByConn = new HashMap<>();
+		List<Repository> repoMembers = federation.getMembers();
+		for(int i=0; i<members.size(); i++) {
+			RepositoryBloomFilter bloomFilter = bloomFilters.get(repoMembers.get(i));
+			if (bloomFilter == null) {
+				bloomFilter = defaultBloomFilter;
+			}
+			bloomFiltersByConn.put(members.get(i), bloomFilter);
+		}
+		com.google.common.base.Function<RepositoryConnection, RepositoryBloomFilter> bloomFilterFunction = Functions.forMap(
```

Comment:
Given that we're Java 8 shop, shouldn't we use java.util.function.Function for this instead?